Fix for JsPointerToString and JsCreateString #6669

dagarxji opened the pull request:
#6669
- Update for JsPointerToString and JsCreateString to ignore input string if length is 0. The function will return an empty string instead of JsErrorNullArgument.
- This passed all unit tests.
- The test included in the JsRTApiTest will fail without these changes.

Tagged as per issue request:
@rhuanjl

dagarxji committed:
Fix for issue #6669

Update for JsPointerToString and JsCreateString to ignore input string if length is 0. The function will return an empty string instead of JsErrorNullArgument.

rhuanjl commented on bin/NativeTests/JsRTApiTest.cpp:
```diff
@@ -165,6 +165,13 @@ namespace JsRTApiTest
+
+		JsValueRef nullFname;
+		REQUIRE(JsCreateString(NULL, 0, &nullFname) == JsNoError);
+
```
Please could you use spaces instead of tabs for the indent (here and other files) - our code base uses spaces consistently.

rhuanjl commented on lib/Jsrt/Core/JsrtCore.cpp:
```diff
@@ -868,10 +868,15 @@ static void CastCopy(const SrcChar* src, DstChar* dst, size_t count)
+	if (length == 0)
+	{
+		content = "";
+	}
```
Please could you bring this in line with the way you've done JsPointerToString below? If length = 0, skip the `PARAM_NOT_NULL(content);` and then use `scriptContext->GetLibrary()->GetEmptyString();` instead of the logic on lines 896 and 898 below.

dagarxji replied:
Oh, I see. I had issues with GetEmptyString() because I was including line 898. Executing 896 and 898 only when length isn't zero fixed this. Thank you.

rhuanjl reviewed (commented):
Thanks for the PR, always good to see a new contributor.

I've added a couple of comments below, but as well as that:
- please could you sign our contribution agreement (add name and username to bottom of this file): https://github.com/chakra-core/ChakraCore/blob/master/ContributionAgreement.md
- please update the copyright banners at the top of the files you've edited, see top of this file for example updated copyright: https://github.com/chakra-core/ChakraCore/blob/master/lib/Runtime/LibraryFunction.h

dagarxji committed:
Update ContributionAgreement.md

dagarxji committed:
Fix for issue #6669 - 2

Update for JsPointerToString and JsCreateString to ignore input string if length is 0. The function will return an empty string instead of JsErrorNullArgument.

dagarxji committed:
Merge branch 'chakracore.git' of https://github.com/dagarxji/ChakraCore into chakracore.git

dagarxji commented:
Well noted. I have updated the files as requested. Hopefully that will remove all of the issues.

rhuanjl commented:
Looks good apart from the test failure. I think it may be because you've put the PERFORM_JSRT_TTD_RECORD_ACTION macro inside the 'if' in one of the functions - can you move that so it's always called then hopefully the test will pass.

dagarxji committed:
Updated JsrtCore.cpp

Moved PERFORM_JSRT_TTD_RECORD_ACTION to execute outside of if statement.

rhuanjl commented on lib/Jsrt/Core/JsrtCore.cpp:
```diff
@@ -887,13 +891,17 @@ CHAKRA_API JsCreateString(
-
         Js::JavascriptString *stringValue = Js::LiteralStringWithPropertyStringPtr::
-            NewFromCString(content, (CharCount)length, scriptContext->GetLibrary());
-
+        NewFromCString(content, (CharCount)length, scriptContext->GetLibrary());
```
These calls should not happen on the length = 0 path, may be the cause of the test failures

rhuanjl reviewed (commented):
Someone else had made a start on trying to do the same issue - though this is still closer to complete - I've looked through and think I've found the likely cause of the test failure.
